{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess, msprime, tskit, os, statistics, pyslim\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import random\n",
    "# import matplotlib.pyplot as plt\n",
    "# import scipy.stats as stats\n",
    "# import pylab\n",
    "import math\n",
    "# import rpy2\n",
    "# import rpy2.robjects as robjects\n",
    "# from rpy2.robjects.packages import importr\n",
    "\n",
    "# energy = importr(\"energy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "Nrun = 100 # number of SIS runs\n",
    "Ne = 6200 \n",
    "Stop = 11 # the stopping criteria\n",
    "SeqLen = 5.0e06 # sequence length\n",
    "MutRate = 1e-9 # mutation rate\n",
    "RecRate = 0 # recombination rate\n",
    "Tau = 50\n",
    "SampleSize = 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For writing input files\n",
    "def WriteInput(ts, Nrun, SampleSize, Ne, Stop, SeqLen, MutRate, RecRate, filename):\n",
    "    filename_tmp = filename + \"1\"\n",
    "    f1 = open(filename_tmp,\"w+\")\n",
    "    f1.write(\"Number of runs = %d\\n\" % Nrun)\n",
    "    f1.write(\"Effective population size = %d\\n\" % Ne)\n",
    "    f1.write(\"Recombination = %.15f\\n\" % RecRate)\n",
    "    f1.write(\"Mutation = %.15f\\n\" % MutRate)\n",
    "    f1.write(\"Length = %f\\n\" % SeqLen)\n",
    "    f1.write(\"Stop = %d\\n\" % Stop)\n",
    "    f1.write(\"#\")\n",
    "    f1.close()\n",
    "    \n",
    "    filename_tmp = filename + \"2\"\n",
    "    f2 = open(filename_tmp,\"w+\")\n",
    "    \n",
    "    # Get distinct haplotypes\n",
    "    geno =  ts.genotype_matrix() \n",
    "    L = len(geno)\n",
    "    f2.write(\"Loci = %d\\n\" % (L+1))\n",
    "    SampleSize = ts.get_sample_size()\n",
    "    f2.write(\"Genes = %d\\n\" % SampleSize)\n",
    "          \n",
    "    # initialise with the first distinct haplotype\n",
    "    haps_1 = [0]\n",
    "    haps_2 = [geno[i][0] for i in range(L)]\n",
    "    haps = np.hstack((haps_1, haps_2))\n",
    "    # the number of this haplotype is 1\n",
    "    #haps.append(1)\n",
    "    haps = np.hstack((haps, [1]))\n",
    "    haps = np.array([haps])\n",
    "    ntype = 1\n",
    "    order = [0]\n",
    "\n",
    "    for j in range(1,SampleSize):\n",
    "        new_1 = [0]\n",
    "        new_2 = [geno[i][j] for i in range(L)]\n",
    "        new = np.hstack((new_1, new_2))\n",
    "        flag = 0 # flag=0: a new distinct haplotype\n",
    "        for k in range(ntype):\n",
    "            if(haps[k][:(L+1)].tolist() == new.tolist()):\n",
    "                haps[k][(L+1)] += 1\n",
    "                flag = 1\n",
    "                order.append(j)\n",
    "                break\n",
    "        if flag == 0:\n",
    "            order.append(ntype)\n",
    "            ntype += 1\n",
    "            #new.append(1)\n",
    "            new = np.hstack((new, [1]))\n",
    "            haps = np.vstack([haps, new]) \n",
    "          \n",
    "    f2.write(\"Distinct Haplotypes = %d\\n\" % ntype)\n",
    "    tot=0\n",
    "    for i in range(ntype):\n",
    "        tot = tot+ np.sum(haps[i,:(L+1)])*haps[i,(L+1)]\n",
    "    stationary = tot/(L*SampleSize)\n",
    "    #stationary = np.sum(haps)/(L*SampleSize)\n",
    "    f2.write(\"Stationary = %.5f\\n\" % (1-stationary))\n",
    "    print(\"stationary = %.5f\\n\" % (1-stationary))  \n",
    "    \n",
    "    f2.write(\"Haplotypes\\n\")\n",
    "    for i in range(haps.shape[0]):\n",
    "        for j in range(haps.shape[1]):\n",
    "            if j == L+1:\n",
    "                f2.write(\"%d\\n\" % haps[i][j])\n",
    "            else:\n",
    "                f2.write(\"%d \" % haps[i][j])\n",
    "    \n",
    "    # Mutation transition rates\n",
    "    f2.write(\"\\nMutation\\n0.0 1.0 0.0 1.0\\n\")\n",
    "             \n",
    "    # Genetic positions\n",
    "    f2.write(\"\\nPositions\\n\")\n",
    "    positions = [0.0]\n",
    "    for s in ts.sites():\n",
    "        positions.append(s.position)\n",
    "    for i in range(len(positions)):\n",
    "        f2.write(\"%.2f \" %(positions[i]))\n",
    "             \n",
    "    f2.write(\"\\n#\")\n",
    "    f2.close()\n",
    "    order = sorted(range(len(order)), key=lambda k: order[k])\n",
    "    return order;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = pyslim.load(\"recipe_50gb.trees\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clade 39 contribute to 3 samples\n",
      "clade 137 contribute to 52 samples\n",
      "clade 153 contribute to 14 samples\n",
      "clade 192 contribute to 1 samples\n",
      "clade 235 contribute to 35 samples\n",
      "clade 258 contribute to 12 samples\n",
      "clade 278 contribute to 8 samples\n",
      "clade 287 contribute to 1 samples\n",
      "clade 296 contribute to 7 samples\n",
      "clade 447 contribute to 19 samples\n",
      "clade 487 contribute to 54 samples\n",
      "clade 513 contribute to 16 samples\n",
      "clade 521 contribute to 20 samples\n",
      "clade 589 contribute to 103 samples\n",
      "clade 612 contribute to 22 samples\n",
      "clade 618 contribute to 31 samples\n",
      "clade 642 contribute to 25 samples\n",
      "clade 666 contribute to 57 samples\n",
      "clade 688 contribute to 28 samples\n",
      "clade 692 contribute to 34 samples\n",
      "clade 738 contribute to 12 samples\n",
      "clade 757 contribute to 9 samples\n",
      "clade 765 contribute to 3 samples\n",
      "clade 788 contribute to 8 samples\n",
      "clade 847 contribute to 12 samples\n",
      "clade 855 contribute to 30 samples\n",
      "clade 990 contribute to 22 samples\n",
      "clade 992 contribute to 24 samples\n",
      "clade 1009 contribute to 34 samples\n",
      "clade 1016 contribute to 33 samples\n",
      "clade 1048 contribute to 3 samples\n",
      "clade 1049 contribute to 20 samples\n",
      "clade 1055 contribute to 31 samples\n",
      "clade 1068 contribute to 7 samples\n",
      "clade 1073 contribute to 45 samples\n",
      "clade 1170 contribute to 6 samples\n",
      "clade 1185 contribute to 55 samples\n",
      "clade 1215 contribute to 32 samples\n",
      "clade 1260 contribute to 24 samples\n",
      "clade 1268 contribute to 16 samples\n",
      "clade 1272 contribute to 10 samples\n",
      "clade 1294 contribute to 11 samples\n",
      "clade 1342 contribute to 4 samples\n",
      "clade 1429 contribute to 22 samples\n",
      "clade 1436 contribute to 11 samples\n",
      "clade 1449 contribute to 25 samples\n",
      "clade 1543 contribute to 23 samples\n",
      "clade 1600 contribute to 9 samples\n",
      "clade 1610 contribute to 71 samples\n",
      "clade 1630 contribute to 6 samples\n",
      "clade 1634 contribute to 15 samples\n",
      "clade 1672 contribute to 56 samples\n",
      "clade 1699 contribute to 67 samples\n",
      "clade 1711 contribute to 86 samples\n",
      "clade 1713 contribute to 38 samples\n",
      "clade 1768 contribute to 30 samples\n",
      "clade 1805 contribute to 21 samples\n",
      "clade 1866 contribute to 36 samples\n",
      "clade 1871 contribute to 11 samples\n",
      "clade 1872 contribute to 27 samples\n",
      "clade 1889 contribute to 16 samples\n",
      "clade 1948 contribute to 35 samples\n",
      "clade 1958 contribute to 43 samples\n",
      "clade 1997 contribute to 25 samples\n",
      "clade 2000 contribute to 50 samples\n",
      "clade 2008 contribute to 19 samples\n",
      "clade 2022 contribute to 23 samples\n",
      "clade 2043 contribute to 19 samples\n",
      "clade 2045 contribute to 57 samples\n",
      "clade 2059 contribute to 31 samples\n",
      "clade 2072 contribute to 2 samples\n",
      "clade 2100 contribute to 44 samples\n",
      "clade 2104 contribute to 4 samples\n",
      "clade 2127 contribute to 38 samples\n",
      "clade 2147 contribute to 64 samples\n",
      "clade 2161 contribute to 13 samples\n",
      "clade 2226 contribute to 18 samples\n",
      "clade 2238 contribute to 17 samples\n",
      "clade 2248 contribute to 57 samples\n",
      "clade 2341 contribute to 9 samples\n",
      "clade 2346 contribute to 14 samples\n",
      "clade 2347 contribute to 16 samples\n",
      "clade 2358 contribute to 17 samples\n",
      "clade 2385 contribute to 21 samples\n",
      "clade 2406 contribute to 24 samples\n",
      "clade 2498 contribute to 8 samples\n",
      "clade 2507 contribute to 4 samples\n",
      "clade 2521 contribute to 9 samples\n",
      "clade 2528 contribute to 12 samples\n",
      "clade 2542 contribute to 1 samples\n",
      "clade 2561 contribute to 7 samples\n",
      "clade 2594 contribute to 102 samples\n",
      "clade 2632 contribute to 52 samples\n",
      "clade 2667 contribute to 46 samples\n",
      "clade 2689 contribute to 21 samples\n",
      "clade 2705 contribute to 8 samples\n",
      "clade 2731 contribute to 7 samples\n",
      "clade 2771 contribute to 10 samples\n",
      "clade 2788 contribute to 34 samples\n",
      "clade 2812 contribute to 19 samples\n",
      "clade 2840 contribute to 9 samples\n",
      "clade 2870 contribute to 71 samples\n",
      "clade 2917 contribute to 14 samples\n",
      "clade 2927 contribute to 20 samples\n",
      "clade 2930 contribute to 19 samples\n",
      "clade 2992 contribute to 25 samples\n",
      "clade 3035 contribute to 28 samples\n",
      "clade 3099 contribute to 16 samples\n",
      "clade 3122 contribute to 6 samples\n",
      "clade 3161 contribute to 2 samples\n",
      "clade 3181 contribute to 20 samples\n",
      "clade 3203 contribute to 67 samples\n",
      "clade 3206 contribute to 56 samples\n",
      "clade 3226 contribute to 44 samples\n",
      "clade 3238 contribute to 4 samples\n",
      "clade 3266 contribute to 50 samples\n",
      "clade 3302 contribute to 38 samples\n",
      "clade 3318 contribute to 41 samples\n",
      "clade 3324 contribute to 54 samples\n",
      "clade 3329 contribute to 6 samples\n",
      "clade 3331 contribute to 49 samples\n",
      "clade 3387 contribute to 29 samples\n",
      "clade 3409 contribute to 16 samples\n",
      "clade 3413 contribute to 5 samples\n",
      "clade 3451 contribute to 38 samples\n",
      "clade 3479 contribute to 149 samples\n",
      "clade 3496 contribute to 5 samples\n",
      "clade 3524 contribute to 44 samples\n",
      "clade 3558 contribute to 1 samples\n",
      "clade 3569 contribute to 15 samples\n",
      "clade 3575 contribute to 42 samples\n",
      "clade 3601 contribute to 43 samples\n",
      "clade 3635 contribute to 10 samples\n",
      "clade 3639 contribute to 2 samples\n",
      "clade 3672 contribute to 39 samples\n",
      "clade 3675 contribute to 6 samples\n",
      "clade 3689 contribute to 9 samples\n",
      "clade 3702 contribute to 40 samples\n",
      "clade 3739 contribute to 5 samples\n",
      "clade 3760 contribute to 32 samples\n",
      "clade 3790 contribute to 57 samples\n",
      "clade 3801 contribute to 26 samples\n",
      "clade 3811 contribute to 13 samples\n",
      "clade 3822 contribute to 9 samples\n",
      "clade 3870 contribute to 14 samples\n",
      "clade 3909 contribute to 2 samples\n",
      "clade 3991 contribute to 62 samples\n",
      "clade 3996 contribute to 1 samples\n",
      "clade 4008 contribute to 59 samples\n",
      "clade 4057 contribute to 18 samples\n",
      "clade 4079 contribute to 34 samples\n",
      "clade 4248 contribute to 17 samples\n",
      "clade 4281 contribute to 24 samples\n",
      "clade 4287 contribute to 11 samples\n",
      "clade 4292 contribute to 9 samples\n",
      "clade 4299 contribute to 38 samples\n",
      "clade 4334 contribute to 73 samples\n",
      "clade 4342 contribute to 94 samples\n",
      "clade 4352 contribute to 18 samples\n",
      "clade 4412 contribute to 8 samples\n",
      "clade 4445 contribute to 9 samples\n",
      "clade 4466 contribute to 87 samples\n",
      "clade 4478 contribute to 15 samples\n",
      "clade 4483 contribute to 33 samples\n",
      "clade 4493 contribute to 1 samples\n",
      "clade 4508 contribute to 31 samples\n",
      "clade 4519 contribute to 13 samples\n",
      "clade 4525 contribute to 26 samples\n",
      "clade 4557 contribute to 22 samples\n",
      "clade 4566 contribute to 50 samples\n",
      "clade 4573 contribute to 19 samples\n",
      "clade 4589 contribute to 4 samples\n",
      "clade 4611 contribute to 28 samples\n",
      "clade 4612 contribute to 56 samples\n",
      "clade 4620 contribute to 92 samples\n",
      "clade 4680 contribute to 34 samples\n",
      "clade 4731 contribute to 20 samples\n",
      "clade 4792 contribute to 39 samples\n",
      "clade 4794 contribute to 4 samples\n",
      "clade 4814 contribute to 53 samples\n",
      "clade 4827 contribute to 49 samples\n",
      "clade 4867 contribute to 47 samples\n",
      "clade 4868 contribute to 10 samples\n",
      "clade 4902 contribute to 3 samples\n",
      "clade 4906 contribute to 13 samples\n",
      "clade 4918 contribute to 17 samples\n",
      "clade 4989 contribute to 73 samples\n",
      "clade 5044 contribute to 18 samples\n",
      "clade 5083 contribute to 34 samples\n",
      "clade 5094 contribute to 28 samples\n",
      "clade 5172 contribute to 5 samples\n",
      "clade 5206 contribute to 18 samples\n",
      "clade 5215 contribute to 38 samples\n",
      "clade 5236 contribute to 10 samples\n",
      "clade 5347 contribute to 10 samples\n",
      "clade 5370 contribute to 57 samples\n",
      "clade 5382 contribute to 14 samples\n",
      "clade 5383 contribute to 8 samples\n",
      "clade 5391 contribute to 8 samples\n",
      "clade 5399 contribute to 14 samples\n",
      "clade 5428 contribute to 80 samples\n",
      "clade 5533 contribute to 19 samples\n",
      "clade 5580 contribute to 3 samples\n",
      "clade 5588 contribute to 27 samples\n",
      "clade 5595 contribute to 71 samples\n",
      "clade 5626 contribute to 23 samples\n",
      "clade 5633 contribute to 18 samples\n",
      "clade 5655 contribute to 8 samples\n",
      "clade 5673 contribute to 27 samples\n",
      "clade 5707 contribute to 58 samples\n",
      "clade 5767 contribute to 7 samples\n",
      "clade 5776 contribute to 16 samples\n",
      "clade 5832 contribute to 11 samples\n",
      "clade 5874 contribute to 6 samples\n",
      "clade 5900 contribute to 55 samples\n",
      "clade 5914 contribute to 14 samples\n",
      "clade 6005 contribute to 4 samples\n",
      "clade 6006 contribute to 68 samples\n",
      "clade 6024 contribute to 18 samples\n",
      "clade 6034 contribute to 27 samples\n",
      "clade 6047 contribute to 13 samples\n",
      "clade 6051 contribute to 13 samples\n",
      "clade 6060 contribute to 12 samples\n",
      "clade 6073 contribute to 15 samples\n",
      "clade 6095 contribute to 5 samples\n",
      "clade 6138 contribute to 5 samples\n",
      "clade 6151 contribute to 17 samples\n",
      "clade 6193 contribute to 6 samples\n",
      "clade 6196 contribute to 3 samples\n",
      "clade 6211 contribute to 4 samples\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clade 6266 contribute to 18 samples\n",
      "clade 6269 contribute to 3 samples\n",
      "clade 6274 contribute to 59 samples\n",
      "clade 6277 contribute to 3 samples\n",
      "clade 6286 contribute to 7 samples\n",
      "clade 6350 contribute to 35 samples\n",
      "clade 6354 contribute to 16 samples\n",
      "clade 6367 contribute to 29 samples\n",
      "clade 6402 contribute to 3 samples\n",
      "clade 6409 contribute to 15 samples\n",
      "clade 6411 contribute to 2 samples\n",
      "clade 6428 contribute to 23 samples\n",
      "clade 6491 contribute to 23 samples\n",
      "clade 6507 contribute to 16 samples\n",
      "clade 6543 contribute to 17 samples\n",
      "clade 6554 contribute to 14 samples\n",
      "clade 6568 contribute to 6 samples\n",
      "clade 6594 contribute to 13 samples\n",
      "clade 6598 contribute to 10 samples\n",
      "clade 6614 contribute to 16 samples\n",
      "clade 6624 contribute to 28 samples\n",
      "clade 6642 contribute to 6 samples\n",
      "clade 6643 contribute to 17 samples\n",
      "clade 6649 contribute to 159 samples\n",
      "clade 6660 contribute to 6 samples\n",
      "clade 6681 contribute to 13 samples\n",
      "clade 6689 contribute to 38 samples\n",
      "clade 6698 contribute to 9 samples\n",
      "clade 6709 contribute to 17 samples\n",
      "clade 6716 contribute to 29 samples\n",
      "clade 6738 contribute to 4 samples\n",
      "clade 6743 contribute to 53 samples\n",
      "clade 6774 contribute to 22 samples\n",
      "clade 6816 contribute to 36 samples\n",
      "clade 6822 contribute to 11 samples\n",
      "clade 6830 contribute to 9 samples\n",
      "clade 6892 contribute to 15 samples\n",
      "clade 6905 contribute to 12 samples\n",
      "clade 6954 contribute to 4 samples\n",
      "clade 6959 contribute to 3 samples\n",
      "clade 6967 contribute to 6 samples\n",
      "clade 7000 contribute to 46 samples\n",
      "clade 7011 contribute to 53 samples\n",
      "clade 7020 contribute to 44 samples\n",
      "clade 7041 contribute to 36 samples\n",
      "clade 7048 contribute to 35 samples\n",
      "clade 7086 contribute to 61 samples\n",
      "clade 7093 contribute to 33 samples\n",
      "clade 7167 contribute to 4 samples\n",
      "clade 7180 contribute to 2 samples\n",
      "clade 7230 contribute to 9 samples\n",
      "clade 7269 contribute to 3 samples\n",
      "clade 7277 contribute to 59 samples\n",
      "clade 7295 contribute to 24 samples\n",
      "clade 7306 contribute to 18 samples\n",
      "clade 7330 contribute to 33 samples\n",
      "clade 7376 contribute to 28 samples\n",
      "clade 7436 contribute to 9 samples\n",
      "clade 7476 contribute to 25 samples\n",
      "clade 7484 contribute to 24 samples\n",
      "clade 7503 contribute to 14 samples\n",
      "clade 7520 contribute to 27 samples\n",
      "clade 7529 contribute to 83 samples\n",
      "clade 7530 contribute to 7 samples\n",
      "clade 7545 contribute to 22 samples\n",
      "clade 7558 contribute to 6 samples\n",
      "clade 7562 contribute to 15 samples\n",
      "clade 7563 contribute to 28 samples\n",
      "clade 7593 contribute to 14 samples\n",
      "clade 7610 contribute to 10 samples\n",
      "clade 7613 contribute to 22 samples\n",
      "clade 7633 contribute to 56 samples\n",
      "clade 7657 contribute to 29 samples\n",
      "clade 7669 contribute to 21 samples\n",
      "clade 7713 contribute to 6 samples\n",
      "clade 7744 contribute to 14 samples\n",
      "clade 7801 contribute to 30 samples\n",
      "clade 7970 contribute to 5 samples\n",
      "clade 7977 contribute to 30 samples\n",
      "clade 7990 contribute to 13 samples\n",
      "clade 8009 contribute to 14 samples\n",
      "clade 8028 contribute to 39 samples\n",
      "clade 8107 contribute to 26 samples\n",
      "clade 8109 contribute to 25 samples\n",
      "clade 8115 contribute to 23 samples\n",
      "clade 8116 contribute to 28 samples\n",
      "clade 8130 contribute to 19 samples\n",
      "clade 8134 contribute to 38 samples\n",
      "clade 8165 contribute to 54 samples\n",
      "clade 8182 contribute to 16 samples\n",
      "clade 8224 contribute to 3 samples\n",
      "clade 8232 contribute to 3 samples\n",
      "clade 8247 contribute to 46 samples\n",
      "clade 8249 contribute to 1 samples\n",
      "clade 8271 contribute to 2 samples\n",
      "clade 8291 contribute to 97 samples\n",
      "clade 8295 contribute to 7 samples\n",
      "clade 8318 contribute to 14 samples\n",
      "clade 8338 contribute to 43 samples\n",
      "clade 8366 contribute to 47 samples\n",
      "clade 8370 contribute to 20 samples\n",
      "clade 8383 contribute to 4 samples\n",
      "clade 8386 contribute to 4 samples\n",
      "clade 8387 contribute to 2 samples\n",
      "clade 8396 contribute to 10 samples\n",
      "clade 8474 contribute to 2 samples\n",
      "clade 8486 contribute to 9 samples\n",
      "clade 8493 contribute to 55 samples\n",
      "clade 8506 contribute to 64 samples\n",
      "clade 8521 contribute to 4 samples\n",
      "clade 8546 contribute to 5 samples\n",
      "clade 8599 contribute to 56 samples\n",
      "clade 8650 contribute to 61 samples\n",
      "clade 8680 contribute to 64 samples\n",
      "clade 8761 contribute to 34 samples\n",
      "clade 8763 contribute to 21 samples\n",
      "clade 8776 contribute to 8 samples\n",
      "clade 8798 contribute to 1 samples\n",
      "clade 8848 contribute to 32 samples\n",
      "clade 8868 contribute to 38 samples\n",
      "clade 8952 contribute to 64 samples\n",
      "clade 8993 contribute to 3 samples\n",
      "clade 9004 contribute to 57 samples\n",
      "clade 9013 contribute to 50 samples\n",
      "clade 9031 contribute to 7 samples\n",
      "clade 9035 contribute to 3 samples\n",
      "clade 9054 contribute to 43 samples\n",
      "clade 9078 contribute to 33 samples\n",
      "clade 9107 contribute to 61 samples\n",
      "clade 9167 contribute to 7 samples\n",
      "clade 9176 contribute to 43 samples\n",
      "clade 9184 contribute to 16 samples\n",
      "clade 9197 contribute to 13 samples\n",
      "clade 9213 contribute to 21 samples\n",
      "clade 9343 contribute to 28 samples\n",
      "clade 9356 contribute to 14 samples\n",
      "clade 9426 contribute to 8 samples\n",
      "clade 9473 contribute to 26 samples\n",
      "clade 9553 contribute to 2 samples\n",
      "clade 9573 contribute to 3 samples\n",
      "clade 9632 contribute to 11 samples\n",
      "clade 9701 contribute to 42 samples\n",
      "clade 9717 contribute to 20 samples\n",
      "clade 9759 contribute to 66 samples\n",
      "clade 9779 contribute to 72 samples\n",
      "clade 9790 contribute to 21 samples\n",
      "clade 9791 contribute to 11 samples\n",
      "clade 9794 contribute to 18 samples\n",
      "clade 9839 contribute to 5 samples\n",
      "clade 9846 contribute to 73 samples\n",
      "clade 9855 contribute to 48 samples\n",
      "clade 9900 contribute to 51 samples\n",
      "clade 9902 contribute to 29 samples\n",
      "clade 9979 contribute to 75 samples\n",
      "clade 10014 contribute to 32 samples\n",
      "clade 10019 contribute to 27 samples\n",
      "clade 10087 contribute to 5 samples\n",
      "clade 10091 contribute to 162 samples\n",
      "clade 10095 contribute to 5 samples\n",
      "clade 10101 contribute to 11 samples\n",
      "clade 10103 contribute to 10 samples\n",
      "clade 10105 contribute to 28 samples\n",
      "clade 10128 contribute to 6 samples\n",
      "clade 10140 contribute to 8 samples\n",
      "clade 10176 contribute to 35 samples\n",
      "clade 10236 contribute to 6 samples\n",
      "clade 10267 contribute to 10 samples\n",
      "clade 10331 contribute to 130 samples\n",
      "clade 10410 contribute to 20 samples\n",
      "clade 10447 contribute to 11 samples\n",
      "clade 10484 contribute to 36 samples\n",
      "clade 10515 contribute to 46 samples\n",
      "clade 10574 contribute to 18 samples\n",
      "clade 10661 contribute to 61 samples\n",
      "clade 10709 contribute to 31 samples\n",
      "clade 10749 contribute to 19 samples\n",
      "clade 10794 contribute to 12 samples\n",
      "clade 10813 contribute to 16 samples\n",
      "clade 10837 contribute to 25 samples\n",
      "clade 10838 contribute to 13 samples\n",
      "clade 10840 contribute to 43 samples\n",
      "clade 10849 contribute to 22 samples\n",
      "clade 10853 contribute to 8 samples\n",
      "clade 10863 contribute to 5 samples\n",
      "clade 10864 contribute to 19 samples\n",
      "clade 10875 contribute to 67 samples\n",
      "clade 10893 contribute to 9 samples\n",
      "clade 10894 contribute to 3 samples\n",
      "clade 10934 contribute to 5 samples\n",
      "clade 10942 contribute to 1 samples\n",
      "clade 10954 contribute to 10 samples\n",
      "clade 10958 contribute to 23 samples\n",
      "clade 11009 contribute to 21 samples\n",
      "clade 11052 contribute to 19 samples\n",
      "clade 11059 contribute to 14 samples\n",
      "clade 11106 contribute to 117 samples\n",
      "clade 11136 contribute to 13 samples\n",
      "clade 11139 contribute to 13 samples\n",
      "clade 11203 contribute to 3 samples\n",
      "clade 11209 contribute to 14 samples\n",
      "clade 11232 contribute to 26 samples\n",
      "clade 11244 contribute to 16 samples\n",
      "clade 11252 contribute to 18 samples\n",
      "clade 11277 contribute to 3 samples\n",
      "clade 11304 contribute to 49 samples\n",
      "clade 11305 contribute to 36 samples\n",
      "clade 11359 contribute to 31 samples\n",
      "clade 11390 contribute to 59 samples\n",
      "clade 11431 contribute to 79 samples\n",
      "clade 11513 contribute to 24 samples\n",
      "clade 11543 contribute to 27 samples\n",
      "clade 11552 contribute to 13 samples\n",
      "clade 11556 contribute to 52 samples\n",
      "clade 11594 contribute to 3 samples\n",
      "clade 11665 contribute to 5 samples\n",
      "clade 11669 contribute to 46 samples\n",
      "clade 11716 contribute to 45 samples\n",
      "clade 11742 contribute to 12 samples\n",
      "clade 11744 contribute to 18 samples\n",
      "clade 11745 contribute to 29 samples\n",
      "clade 11789 contribute to 26 samples\n",
      "clade 11841 contribute to 10 samples\n",
      "clade 11858 contribute to 12 samples\n",
      "clade 11862 contribute to 3 samples\n",
      "clade 11873 contribute to 25 samples\n",
      "clade 11932 contribute to 34 samples\n",
      "clade 11948 contribute to 26 samples\n",
      "clade 11954 contribute to 26 samples\n",
      "clade 11962 contribute to 46 samples\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clade 11999 contribute to 9 samples\n",
      "clade 12017 contribute to 30 samples\n",
      "clade 12026 contribute to 16 samples\n",
      "clade 12079 contribute to 159 samples\n",
      "clade 12120 contribute to 8 samples\n",
      "clade 12170 contribute to 23 samples\n",
      "clade 12173 contribute to 37 samples\n",
      "clade 12229 contribute to 23 samples\n",
      "clade 12312 contribute to 14 samples\n",
      "clade 12327 contribute to 77 samples\n",
      "clade 12330 contribute to 3 samples\n",
      "clade 12387 contribute to 2 samples\n"
     ]
    }
   ],
   "source": [
    "tot = 0\n",
    "for c in range(Ne*2):\n",
    "    clade1 = c\n",
    "    count = 0\n",
    "    ts = tree.at(100)\n",
    "    for s in range(Ne*2, Ne*4):\n",
    "        if ts.is_descendant(s, clade1):\n",
    "            count = count + 1\n",
    "    if count>0:\n",
    "        tot = tot + count\n",
    "        print(\"clade %d contribute to %d samples\" %(c, count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[14146, 14623, 14641, 14983, 17896, 19789, 19958, 20118, 23607, 24489]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roots = [1272] #1272, 39\n",
    "s1 = []\n",
    "\n",
    "ts = tree.at(100)\n",
    "for s in range(Ne*2, Ne*4):\n",
    "    if ts.is_descendant(s, roots[0]):\n",
    "        s1.append(s)\n",
    "cases = s1\n",
    "cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([15959, 20439, 23864, 23780, 23564, 13906, 23426, 24470, 12683,\n",
       "       21548])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.random.seed(19)\n",
    "rest = [x for x in range(Ne*2, Ne*4) if x not in cases]\n",
    "control = np.random.choice(rest, size=(SampleSize - len(cases)), replace=False)\n",
    "control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(475, 20)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tree_re = tree.recapitate(recombination_rate=0, Ne=Ne, random_seed=1)\n",
    "com = np.hstack([cases, control])\n",
    "tree_com = tree_re.simplify(samples=com, keep_unary=True)\n",
    "tree_com = msprime.mutate(tree_com, rate=MutRate, random_seed=1, keep=True)\n",
    "geno_com = tree_com.genotype_matrix()\n",
    "geno_com.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stationary = 0.58137\n",
      "\n"
     ]
    }
   ],
   "source": [
    "filename = \"infile\"\n",
    "order_com = WriteInput(tree_com, Nrun, SampleSize, Ne, Stop, SeqLen, MutRate, RecRate, filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
